### ISSUE TITLE
Post-dominator based scope calculation insufficient

### ISSUE THREAD

**Issue by nomeata** [nomeata/incredible#54]
@lohner pointed me to a bug today, and I tracked it down to a insufficiency in the scope calculation.

This is the failure:

![unfication-regression1](https://cloud.githubusercontent.com/assets/148037/10259162/749d1d52-6965-11e5-88bc-44f738fa45ee.png)

It should say A(c₄)∨B(c₄), but it dropped the dependency on c. The problem was *not* the unification, but the scope calculation in the presence of open ports.

Here is what it does now: Every block that has *no* successor gets an implicit edge to a fake exit note. If we’d remove the left edge, then the ∃-elim block would have no such edge, so it would get one, then the ∨ block would be post-dominated by ∃’s local input, and A and C would be allowed to depend on c₄, as observed by Denis.

With this connection, now both blocks do have a successor (the other block), so no edge to the exit is added, and thus no post dominators are found at all!

This points to the work-around: Add a block after ∃ that has no successors, as in

![unfication-regression2](https://cloud.githubusercontent.com/assets/148037/10259215/005f28c6-6966-11e5-9345-8b730202937e.png)

Now the scope algorithm works as intended and unification does as well!

Drawing a line to the fake exit from *all* open ports would also be too harsh.

So we need a better idea about how to define scopes in the presence of open ports.

**Comment by nomeata** [comment]
This should work. Given a block `b` with an assumption `p` opening a scope.
 * From the `b` forward (i.e. follow all outgoing edges that are not `consumedBy p`), mark every block, including `b`. These are definitely not in the scope.
 * From this set, trace backwards following all inputs (excluding `p`). These are also not in the scope.
 * The complement of the union of these two sets are the scope.

Optimization: Restrict the result to the strongly connected component of `b` (in the undirected graph).

The same algorithm can be used to fix #3: The `consumedBy` port defines a scope. Check that all edges from the local hypothesis end up in the scope. Or, alternatively, define the sets above not only on blocks, but also on ports: Then simply check that the local hypothesis is in the scope.

For #3, we also want to know a path that witnesses the problem. In this case, do not produce sets but rather maps from nodes to the length of the shortest path to the start, by doing a breadth first traversal. Then one can find a shortest bad path by going the other way, always to the node with the lowest number.

**Comment by nomeata** [comment]
Implemented it now. I ended up implementing 

 * From the `b` forward, disregarding outputs that are local hypotheses, mark every block, including `b`. These are definitely not in the scope.
 * From this set, trace backwards following all inputs (excluding `p`). These are also not in the scope.
 * Subtract this set from all blocks connected to `p` in any way.

I hope that ignoring all local hypotheses here is not ignoring too much, but this way the test cases pass :-)
